### PR TITLE
model cryption in memory and data cryption in file

### PR DESCRIPTION
**Pull request by chenwhql**

### CODE REVIEW

File: paddle/fluid/framework/cryption.cc [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  PADDLE_ENFORCE(WBAES_OK == result, "WBAES init file failed.");
+
+  result = wbaes_decrypt_file(encryptFilePath.data(), decryptFilePath.data(),
+                              block_size);
+  PADDLE_ENFORCE(WBAES_OK == result, "WBAES encrypt file failed.");
+
+  return;
+}
+
+std::string ConvertHexString(const char* buf, int len) {
```
Comment (liupluswei) [inline paddle/fluid/framework/cryption.cc]: is this function only used for debug?

File: paddle/fluid/framework/cryption.cc [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+  // Encrypt
+  result = wbaes_init(encrypt_key_path_, NULL);
+  PADDLE_ENFORCE(WBAES_OK == result, "WBAES init file failed.");
+
+  result = wbaes_encrypt_file(inputFilePath.data(), encryptFilePath.data(),
+                              block_size);
+  PADDLE_ENFORCE(WBAES_OK == result, "WBAES encrypt file failed.");
+
+  return;
```
Comment (panyx0718) [inline paddle/fluid/framework/cryption.cc]: no need

**Review by panyx0718 — DISMISSED** [review]

File: paddle/fluid/pybind/pybind.cc [hunk 3]
```diff
@@ -1348,6 +1349,35 @@ All parameter, weight, gradient are variables in Paddle.
         self.Run(fetch_tensors, fetched_var_name);
       });
 
+  py::class_<framework::Cryption>(m, "Cryption")
+      .def_static(
+          "get_cryptor",
+          []() {
+            std::unique_ptr<Cryption> cryptor =
```
Comment (liupluswei) [inline paddle/fluid/pybind/pybind.cc]: just curious, can we just return Cryption::GetCryptorInstance()?

File: paddle/fluid/framework/cryption_test.cc [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+
+  fout.open(path, std::ios::binary);
+  fout.write(buf, len);
+
+  fout.close();
+  return;
+}
+
+TEST(Cryption, OnceInitTest) {
+  Cryption& c1 = *Cryption::GetCryptorInstance();
```
Comment (liupluswei) [inline paddle/fluid/framework/cryption_test.cc]: what's the purpose of this unittest, c1 & c2 looks like the same pointer?

File: python/paddle/fluid/tests/unittests/test_cryption.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    def test_cryption_in_memory_divisible_16(self):
+        input_str = str.encode("0123456789abcdef0123456789abcdef")
+
+        # get encrypt len
+        input_str, crypt_len = self._align_bytes_to_16(input_str)
+
+        cryptor = core.Cryption.get_cryptor()
+        encrypt_str = cryptor.encrypt_in_memory(input_str, crypt_len)
+        decrypt_str = cryptor.decrypt_in_memory(encrypt_str, crypt_len)
+        self.assertEqual(input_str, decrypt_str)
```
Comment (liupluswei) [inline python/paddle/fluid/tests/unittests/test_cryption.py]: this is not enough.  I suppose you should remove the padding part after decryption and compare it with the original raw input, right?

File: paddle/fluid/framework/cryption.cc (hunk 1 above)
Comment (chenwhql) [reply paddle/fluid/framework/cryption.cc]: Done, deleted

File: paddle/fluid/framework/cryption.cc (hunk 2 above)
Comment (chenwhql) [reply paddle/fluid/framework/cryption.cc]: Done, deleted

File: paddle/fluid/pybind/pybind.cc (hunk 3 above)
Comment (chenwhql) [reply paddle/fluid/pybind/pybind.cc]: it also works, changed to it

File: paddle/fluid/framework/cryption_test.cc (hunk 4 above)
Comment (chenwhql) [reply paddle/fluid/framework/cryption_test.cc]: Done, deleted

File: python/paddle/fluid/tests/unittests/test_cryption.py (hunk 5 above)
Comment (chenwhql) [reply python/paddle/fluid/tests/unittests/test_cryption.py]: Right, fixed

File: paddle/fluid/framework/cryption.cc [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include <memory>
+#include <string>
+
+#include "paddle/fluid/framework/cryption.h"
```
Comment (chengduoZH) [inline paddle/fluid/framework/cryption.cc]: I think that `cryption` should be placed in `paddle/fluid/contrib` but not `paddle/fluid/framework`.

Comment (chenwhql) [reply paddle/fluid/framework/cryption.cc]: Done

File: paddle/fluid/contrib/cryption.cc [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+Cryption::Cryption() {
+  // Init
+  CreateKeyInFile();
+}
+
+void Cryption::CreateKeyInFile() {
+  int result = WBAES_OK;
+
+  result = wbaes_create_key_in_file(key_string_, encrypt_key_path_,
+                                    decrypt_key_path_);
```
Comment (chengduoZH) [inline paddle/fluid/contrib/cryption.cc]: => `int result = wbaes_create_key_in_file...`

File: paddle/fluid/contrib/cryption.cc [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  return &cryptor;
+}
+
+std::string Cryption::EncryptInMemory(const char* inputStr,
+                                      const size_t strLen) {
+  int result = WBAES_OK;
+
+  // Input Check
+  PADDLE_ENFORCE_NOT_NULL(inputStr, "The input string is null.");
+  PADDLE_ENFORCE(0 != strLen,
```
Comment (chengduoZH) [inline paddle/fluid/contrib/cryption.cc]: => `PADDLE_ENFORCE_NE`

File: paddle/fluid/contrib/cryption.cc [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+                 "The length of the input string should not be equal to 0.");
+  PADDLE_ENFORCE(
```
Comment (chengduoZH) [inline paddle/fluid/contrib/cryption.cc]: =>`PADDLE_ENFORCE_EQ`

File: paddle/fluid/contrib/cryption.h [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  // a import factor for file encryption/decryption
+  //   the encryption/decryption block size should be equal for a file
+  const int block_size = 4096;
+
+  // the variables used by encryption/decryption in file
+  const char* encrypt_key_path_ = "./encrypt.key";
+  const char* decrypt_key_path_ = "./decrypt.key";
+
+  std::unique_ptr<char> encrypt_text_;
+  std::unique_ptr<char> decrypt_text_;
```
Comment (chengduoZH) [inline paddle/fluid/contrib/cryption.h]: `std::unique_ptr<char>` => `std::unique_ptr<char[]>`

File: paddle/fluid/contrib/cryption.h [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+  // used to generate encryption key and decryption key
+  const char* key_string_ = "0123456789abcdef";
+
+  // a import factor for file encryption/decryption
+  //   the encryption/decryption block size should be equal for a file
+  const int block_size = 4096;
+
+  // the variables used by encryption/decryption in file
+  const char* encrypt_key_path_ = "./encrypt.key";
+  const char* decrypt_key_path_ = "./decrypt.key";
```
Comment (chengduoZH) [inline paddle/fluid/contrib/cryption.h:59]: Now that the above variable is `const`, maybe it is not appropriate that putting them in `Cryption `.

File: paddle/fluid/contrib/cryption_test.cc [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+  EXPECT_EQ(inputStr.substr(0, origLen), decryptStr.substr(0, origLen));
+}
+
+TEST(Cryption, CryptWithFile) {
+  Cryption& c = *Cryption::GetCryptorInstance();
+  const char* inputStr = "0123456789abcdef0123456789abcdef";
+  int strLen = strlen(inputStr);
+  char* inputStrCopy = new char[strLen];
+  char* encryptStr = new char[strLen];
+  char* decryptStr = new char[strLen];
```
Comment (chengduoZH) [inline paddle/fluid/contrib/cryption_test.cc]: You can use smart point here, but not use `new` and `delete`.

File: paddle/fluid/contrib/cryption.cc (hunk 9 above)
Comment (chenwhql) [reply paddle/fluid/contrib/cryption.cc]: Done

File: paddle/fluid/contrib/cryption.cc (hunk 7 above)
Comment (chenwhql) [reply paddle/fluid/contrib/cryption.cc]: Done

File: paddle/fluid/contrib/cryption_test.cc [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+void writeFile(const std::string& path, const char* buf, size_t len) {
+  std::ofstream fout;
+
+  fout.open(path, std::ios::binary);
+  fout.write(buf, len);
+
+  fout.close();
+  return;
```
Comment (chengduoZH) [inline paddle/fluid/contrib/cryption_test.cc]: remove ling 44.

File: paddle/fluid/contrib/cryption_test.cc [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+void readFile(const std::string& path, char* buf, size_t len) {
+  std::ifstream fin;
+
+  fin.open(path, std::ios::binary);
+  fin.read(buf, len);
+
+  buf[len] = '\0';
+
+  fin.close();
+  return;
```
Comment (chengduoZH) [inline paddle/fluid/contrib/cryption_test.cc]: remove line 34.

File: paddle/fluid/contrib/cryption_test.cc (hunk 13 above)
Comment (chenwhql) [reply paddle/fluid/contrib/cryption_test.cc]: Done

File: paddle/fluid/contrib/cryption_test.cc (hunk 14 above)
Comment (chenwhql) [reply paddle/fluid/contrib/cryption_test.cc]: Done

File: paddle/fluid/pybind/CMakeLists.txt [hunk 15]
```diff
@@ -1,6 +1,12 @@
...
-  tracer analysis_predictor imperative_profiler)
+if(NOT WITH_WBAES)
+  set(PYBIND_DEPS pybind python proto_desc memory executor async_executor fleet_wrapper prune
+    feed_fetch_method pass_builder parallel_executor profiler layer scope_pool
+    tracer analysis_predictor imperative_profiler)
+else()
+  set(PYBIND_DEPS pybind python proto_desc memory executor async_executor fleet_wrapper prune
+    feed_fetch_method pass_builder parallel_executor profiler layer scope_pool
+    tracer analysis_predictor imperative_profiler cryption)
+endif()
```
Comment (luotao1) [inline paddle/fluid/pybind/CMakeLists.txt:7]: 这里if else基本都重复。建议如下更加干净
```
set(PYBIND_DEPS pybind python proto_desc xxxx)
if (WITH_WBAES)
  set(PYBIND_DEPS ${PYBIND_DEPS} cryption)
endif()
```

Comment (chenwhql) [reply paddle/fluid/pybind/CMakeLists.txt:7]: Done

**Pull request closed without merging**